QWeb/internal/xhr.py

```diff
@@ -16,12 +16,256 @@
+    jqActive: jqActive
+  };
+})(arguments[0]);
+"""
+
+# Separate spinner probe (selectors configurable)
+JS_IS_SPINNER_BUSY = r"""
+return (function (selectors) {
+  function visible(el){
+    if (!el) return false;
+    var cs = getComputedStyle(el);
+    if (cs.visibility==='hidden' || cs.display==='none' || cs.opacity==='0') return false;
+    var r = el.getBoundingClientRect();
+    return r.width>0 && r.height>0;
+  }
+  try {
+    var sels = Array.isArray(selectors) ? selectors : [];
+    for (var i=0;i<sels.length;i++){
+      var nodes = document.querySelectorAll(sels[i]);
+      for (var j=0;j<nodes.length;j++){
+        if (visible(nodes[j])) return true;
+      }
+    }
+    return false;
+  } catch (e) {
+    return false; // if probe fails, treat as not-busy (avoid blocking)
+  }
+})(arguments[0]);
+"""
 
 
 def setup_xhr_monitor() -> bool:
+    try:
+        ok = javascript.execute_javascript(JS_INSTALL_MONITOR)
+        if ok is not True:
+            logger.debug(f"setup_xhr_monitor returned non-True: {ok}")
+        return bool(ok)
+    except JavascriptException as e:
+        logger.debug(f"setup_xhr_monitor failed: {e}")
+        raise QWebDriverError(e)  # pylint: disable=W0707
+
+
+def get_light_status(quiet_ms: int = 400) -> Optional[dict]:
+    """Full status dict. None on failure."""
+    try:
+        st = javascript.execute_javascript(JS_STATUS_LITE, quiet_ms)
+        if isinstance(st, dict):
+            return st
+        logger.debug(f"get_light_status: unexpected return {type(st)}")
+        return None
+    except JavascriptException as e:
+        logger.debug(f"get_light_status failed: {e}")
+        return None
+
+
+def is_spinner_busy(selectors: Optional[list[str]] = None) -> Optional[bool]:
+    """Return True/False if spinner appears visible; None on probe failure."""
+    try:
+        busy = javascript.execute_javascript(JS_IS_SPINNER_BUSY, selectors)
+        # Selenium returns a JS boolean → Python bool; still validate
+        if isinstance(busy, bool):
+            return busy
+        logger.debug(f"is_spinner_busy: unexpected return {type(busy)}")
+        return None
+    except JavascriptException as e:
+        logger.debug(f"is_spinner_busy failed: {e}")
+        return None
+
+
+def _parse_spinner_selectors() -> Optional[list[str]]:
+    raw = config.get_config("SpinnerCSS")
+    if raw is None:
+        return None
+
+    # Normalize common "empty" markers to None
+    if isinstance(raw, str) and raw.strip().lower() in ("", "none", "null", "false", "off"):
+        return None
+
+    # Otherwise interpret as comma-separated CSS selectors
+    if isinstance(raw, str):
+        return [s.strip() for s in raw.split(",") if s.strip()]
+
+    return None
+
+
+def wait_xhr(timeout: float = 15.0,
+             poll_interval: float = 0.1) -> None:
+    """
+    Order: readyState -> network idle -> spinner gone -> DOM quiet (bounded).
+    - `quiet_ms`: quiet window needed to call DOM "settled"
```

Comment:
Could you add the constants and their effects on the behavior to the docstring?
Then when hovering over the function call somewhere else in an IDE you can get a better grasp of the func without traversing to this source file